### ISSUE TITLE
Add simple Health-Check cappability

### ISSUE THREAD

**Issue by rhuss** [rhuss/jolokia#271]
A new configuration option `healthChecks` should be added which can carry a referenced to a JSON file for defining simple health checks. 

Simple mean that no big logic should be added. If more logic is needed then an implementation of https://github.com/rhuss/jolokia-extra/issues/1 should be considered.

A health-check is a reference to JMX READ or EXEC for getting the value of an MBean's attribute or operation. This value is fetched on behalf of a request and compared against a given value, defined also in the health checks. The health-check has also an option to specify how the value should be evaluated to yield an `OK` or `NOK`. If an exception occurs this is translated to `ERROR`

Mutiple health-checks can be defined. The overall result to the health-check endpoint is `OK` when all health-checks yield ok, or `NOK` if not.

The protocol will be enhanced with a new command 'health' (on the same footprint as 'read', 'write', 'list', ...). It yields an error if no health-checks are provided during startup. 

The health-check configuration also can contain an IP and a Port. If given, the agent will open a new port on the given ip/config (JVM agent only), serving **only** the health check endpoint. This can be used when the regular port is highly secured e.g. with SSL certificates, but still a simple mean for health checks should be available (with no or only lightweight authentication). Checks also carry one or more types which could be selected by a query parameter.

Here are some possible examples for a configuration and a corresponding JSON response:

```json
{
  "host": "127.0.0.1",
  "port": "8779"
  "checks": [
    {
      "name": "Heap Memory Usage"
      "types": [ "liveness" ]
      "request": {
        "mbean": "java.lang:type=Memory",
        "attribute": "HeapMemoryUsage",
        "path": "used"
      }
      "rule": "${value} < 1232896"
    },
    {
      "name": "Thread Number",
      "types": [ "readiness", "liveness" ]
      "request": {
        "mbean": "java.lang:type=Threading"
        "attribute": "ThreadCount"
        "rule": "${value} > 10"
      }
    }
  ]
}
```

and the response could be

```json
{
    "result": "NOK",
    "checks": [
      {
        "name": "Heap Memory Usage",
        "result": "OK",
        "value": "567803"
      },
      {
        "name": "Thread Number",
        "result": "NOK",
        "value": "5"
      }
    ]
}
```

**Comment by jstrachan** [comment]
LGTM. 

It'd be nice to detect existing MBean health checks using naming conventions maybe. e.g. if the mbean name is `*HealthCheck` and the method is, say, `check()` for DropWizard: http://metrics.dropwizard.io/3.1.0/manual/healthchecks/

Am not sure if spring boot health checks are available over JMX http://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-endpoints.html#production-ready-health seems they are only visible inside a spring ApplicationContext OOTB; so we could define an MBean to check them all I guess?

**Comment by rhuss** [comment]
Yeah, thats a good idea to automatically pickup custom health checks registered with MBeans. On startup one could do an initial scan for MBeans with a certain name pattern, and then register for JMX notifications when new MBean gets registered and update this list accordingly (there is such a notification which I already use).

That way we could defined health checks from the outside (via JSON) and the inside (via MBeans). And it would fit nicely in the Jolokia model since its not really intrusive.

For spring boot health checks we could code a starter jar which registers MBean according to our naming scheme and then delegate to the spring healthchecks beans. This way we can keep Jolokia spring-boot agnostic but still be able to include spring boot health checks. The alternative would be to scan e.g. in a ServerDetector for spring boot healtchecks beans and the register the MBeans transparently, but it could be that this scan could be too early, before spring boot initialization. But well, this is only the second step, lets kick off the health checks first.

Perfect, thanks a lot for idea. think we are now finally ready for health checking ;-)

**Comment by jstrachan** [comment]
Agreed. BTW we could query the health check MBeans whenever the health check endpoint is invoked? It keeps the code simpler & avoids a timer/poller? Its not like querying mbeans is particularly slow or anything and its no biggie if a health check doesn't respond for a couple of ms.

For Spring health checks we could have a single MBean that matches the jolokia naming convention which just invokes all the spring health checks it can find (and aggregates all the results?). Then its nice and simple and just 1 MBean class? Doing it from the outside of the users app would rock; but I can imagine it could be tricky from a class loading perspective maybe; though I guess its only 1 method thats being invoked on all the health checks.

**Comment by rhuss** [comment]
I'm a bit afraid that querying is quite expensive even when using a pattern (when I looked last, it was an iteration over the list of all registered MBeans which can get quite large in servers). But you are right, we should start simple. using the regisration notifications is not that difficult, but we can do it later if we have some numbers (also, we could do a full query from time to time, too to avoid stale caches).

I think, too, that a single spring-boot proxy MBean would be sufficient. There could be indeed classloading issues when not accessing the spring-boot hc via JMX. in the worst case we could go over the HTTP endpoint. Let's see ;-)

**Comment by jstrachan** [comment]
Sounds good ;). I'd be totally happy with a slow health check endpoint (due to the querying) on a massive server that took a few seconds to return (so long as the user doesn't set the polling time to 1 second ;)

**Comment by rhuss** [comment]
I'm not so much worried about the response times but on the effect the query has on the server performance. E.g. for a prometheus scan it takes up to 100% CPU for the query for some time.

**Comment by jstrachan** [comment]
ah gotcha! I guess doing a little registration notification might be a good idea then ;)
